src/qibolab/pulses/shape.py

```diff
@@ -79,94 +81,44 @@ def demodulate(
     return np.sqrt(2) * np.einsum("ijt,jt->it", demod, modulated)
 
 
-class ShapeInitError(RuntimeError):
-    """Error raised when a pulse has not been fully defined."""
-
-    default_msg = "PulseShape attribute pulse must be initialised in order to be able to generate pulse waveforms"
-
-    def __init__(self, msg=None, *args):
-        if msg is None:
-            msg = self.default_msg
-        super().__init__(msg, *args)
-
-
-class PulseShape(ABC):
+class Shape(ABC):
     """Pulse envelopes.
 
     Generates both i (in-phase) and q (quadrature) components.
     """
 
-    pulse = None
-    """Pulse (Pulse): the pulse associated with it.
-
-    Its parameters are used to generate pulse waveforms.
-    """
-
     @abstractmethod
-    def envelope_waveform_i(
-        self, sampling_rate=SAMPLING_RATE
-    ) -> Waveform:  # pragma: no cover
-        raise NotImplementedError
+    def i(self, times: Times) -> Waveform:
+        """In-phase envelope."""
 
     @abstractmethod
-    def envelope_waveform_q(
-        self, sampling_rate=SAMPLING_RATE
-    ) -> Waveform:  # pragma: no cover
-        raise NotImplementedError
+    def q(self, times: Times) -> Waveform:
+        """Quadrature envelope."""
 
-    def envelope_waveforms(self, sampling_rate=SAMPLING_RATE):
-        """A tuple with the i and q envelope waveforms of the pulse."""
+    def envelopes(self, times: Times) -> IqWaveform:
+        """Stacked i and q envelope waveforms of the pulse."""
+        return np.array(self.i(times), self.q(times))
 
-        return (
-            self.envelope_waveform_i(sampling_rate),
-            self.envelope_waveform_q(sampling_rate),
-        )
 
-    def __eq__(self, item) -> bool:
-        """Overloads == operator."""
-        return isinstance(item, type(self))
+@dataclass(frozen=True)
+class Rectangular(Shape):
+    """Rectangular envelope."""
 
-    @staticmethod
-    def eval(value: str) -> "PulseShape":
-        """Deserialize string representation.
+    amplitude: float
```

Comment:
I'm planning to drop this, since it will be the same for all the shapes, and it could just be plugged in by multiplying the final envelope.

Reply:
> could just be plugged in by multiplying the final envelope.

Where would this happen?

Reply:
That's a good question, since the functions are directly called all over the place in the drivers.

There are two options:
1. keep the information in the shape, but make some wrappers in the base class
2. retain the information only in the `Pulse`, and have wrappers in there

Option 1. is actually more decoupled, but there is the unpleasant `@dataclass` inheritance issue to face. Option 2. is reasonable, but you are making `Shape` usage internal and less transparent.

What's your opinion? @hay-k @stavros11

Repeating for each shape is not an option, of course :P

Reply:
I believe my thinking is more along the lines of the second option. We could think of shape as being normalized, so it describes the outline of the pulse, but not the "vertical" size. In a sense, it will describe the outline qualitatively but not quantitatively. In that case the concept of an amplitude becomes meaningless for a shape, and it could become a property of a pulse, and it will act as a scaling factor of given shape, to scale all samples uniformly. We could consider naming it something else than an amplitude as well.

Reply:
Done in 608bb8d